File: Week_02-哈希表-集合-树-图/145.二叉树的后序遍历.py
#
# @lc app=leetcode.cn id=145 lang=python
#
# [145] 二叉树的后序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        '''
        #方法1:递归
        ans = list()
        def dfs(root):
            if not root:
                return
            dfs(root.left)
            dfs(root.right)
            ans.append(root.val)
        dfs(root)
        return ans
        '''
        #方法2:迭代
        ans = list()
        stack = list()
        prev = None
        while root or stack:
            #处理左节点，连续向左
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop() #弹出当前根节点
            #处理右节点
            if root.right and root.right != prev:  #如果右节点不空且没有被访问过，就把当前根节点再次入栈
                stack.append(root)
                root = root.right
            else:
                #处理根节点
                ans.append(root.val)
                #记录刚访问过的根节点
                prev = root
                root = None
            return ans

# @lc code=end

File: Week_02-哈希表-集合-树-图/145.二叉树的后序遍历.cpp
/*
 * @lc app=leetcode.cn id=145 lang=cpp
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
#if 0
    //递归法
    vector<int> ans;
    vector<int> postorderTraversal(TreeNode* root) {
        if(root != NULL) {
            postorderTraversal(root -> left);
            postorderTraversal(root -> right);
            ans.push_back(root -> val);
        }
        return ans;
    }
#endif

#if 1
    /*
    后序遍历用前序遍历（#145题）相似的方法，后序遍历与前序遍历对称，遍历完成后把结果进行逆序处理
    伪代码：
    枚举更新根节点，来自栈中暂存的各级左子树（前序是右子树）
        枚举更新根节点，来自上次循环中根节点的右子树（前序是左子树）
            输出根节点的值；
            栈中暂存左子树（前序是右子树）
        根节点指向右子树（前序是左子树）
    根节点指向栈顶暂存的左子树（前序是右子树），弹栈顶；
    结果序列逆序
    */
    vector<int> postorderTraversal(TreeNode* root) {

        stack<TreeNode*> S;
        vector<int> v;
        TreeNode* rt = root;
        while(rt || !S.empty()){
            while(rt){
                S.push(rt->left);
                v.push_back(rt->val);
                rt = rt->right;
            }
            rt = S.top();S.pop();
        }
        reverse(v.begin(), v.end());
        return v;      
    }
#endif
};
// @lc code=end

